internal/ratelimit/map.go

```diff
@@ -0,0 +1,64 @@
+package ratelimit
+
+import (
+	"net/http"
+	"time"
+)
+
+// Map maps categories to rate limit deadlines.
+//
+// A rate limit is in effect for a given category if either the category's
+// deadline or the deadline for the special CategoryAll has not yet expired.
+//
+// Use IsRateLimited to check whether a category is rate-limited.
+type Map map[Category]Deadline
+
+// IsRateLimited returns true if the category is currently rate limited.
+func (m Map) IsRateLimited(c Category) bool {
+	return m.isRateLimited(c, time.Now())
+}
+
+func (m Map) isRateLimited(c Category, now time.Time) bool {
+	return m.Deadline(c).After(Deadline(now))
+}
+
+// Deadline returns the deadline when the rate limit for the given category or
+// the special CategoryAll expire, whichever is furthest into the future.
+func (m Map) Deadline(c Category) Deadline {
+	categoryDeadline := m[c]
+	allDeadline := m[CategoryAll]
+	if categoryDeadline.After(allDeadline) {
+		return categoryDeadline
+	}
+	return allDeadline
+}
+
+// Merge merges the other map into m.
+//
+// If a category appears in both maps, the deadline that is furthest into the
+// future is preserved.
+func (m Map) Merge(other Map) {
+	for c, d := range other {
+		if d.After(m[c]) {
+			m[c] = d
+		}
+	}
+}
+
+// FromResponse returns a rate limit map from an HTTP response.
+func FromResponse(r *http.Response) Map {
+	return fromResponse(r, time.Now())
+}
+
+func fromResponse(r *http.Response, now time.Time) (m Map) {
+	s := r.Header.Get("X-Sentry-Rate-Limits")
+	if s != "" {
+		return parseXSentryRateLimits(s, now)
```

Comment:
Good question.

This file is all about the `Map` type, which is essentially the most important exported type from this package, since it connects together all the rest. It could be the *only* exported type if we used `time.Time` for deadlines and `string` for categories.

`FromResponse` is the "constructor" for `Map` (oh, and I just spotted something that went in accidentally... signature should be `func fromResponse(r *http.Response, now time.Time) Map`).

The constructor is often organized just like the methods for the given type, and type, methods and constructors are grouped together in `go doc` output:

```console
$ go doc ./internal/ratelimit
package ratelimit // import "github.com/getsentry/sentry-go/internal/ratelimit"

Package ratelimit provides tools to work with rate limits imposed by
Sentry's data ingestion pipeline.

type Category string
    const CategoryAll Category = "" ...
type Deadline time.Time
type Map map[Category]Deadline
    func FromResponse(r *http.Response) Map
```

Then there are two distinct parts for rate limiting:

- The old stuff, `retry_after.go`
- The new stuff, `rate_limits.go` => maybe should be named `x_sentry_rate_limits.go` but that just looked unnecessarily long. I reckon a file `ratelimit/rate_limits.go` does look like the most important thing in a `ratelimit/` directory... 😕 

So all files in a directory are part of the same package, there is nothing uncommon with cross-file references. Similarly, at the root of the repo every file depend on code defined in other files, so at the extreme we'd end up with one file per directory with all of the code -- the compiler wouldn't care.

With all that explanation -- what do you think? Would it be better to rename `rate_limits.go` => `x_sentry_rate_limits.go`?
Moving the constructor of the type away into another file is not common, but I am open to hear your input and make accommodating changes.